lib/ProductOpener/Auth.pm

```diff
@@ -0,0 +1,878 @@
+
+A workable instance of OIDC::Lite::Client::WebServer.
+
+=cut
+
+sub _get_client () {
+	if ($client) {
+		return $client;
+	}
+
+	get_oidc_configuration();
+	$client = OIDC::Lite::Client::WebServer->new(
+		id => $oidc_options{client_id},
+		secret => $oidc_options{client_secret},
+		authorize_uri => $oidc_configuration->{authorization_endpoint},
+		access_token_uri => $oidc_configuration->{token_endpoint},
+	);
+	return $client;
+}
+
+=head2 get_oidc_configuration( )
+
+Ensures that OIDC (OpenID Connect) is discovered and configured.
+
+If OIDC is already discovered, the function returns without doing anything.
+
+Otherwise, it sends a discovery request to the OIDC endpoint and loads the discovery document.
+If successful, it updates the OIDC options with the JWKS (JSON Web Key Set) configuration.
+
+=head3 Arguments
+
+None.
+
+=head3 Return values
+
+None.
+
+=cut
+
+sub get_oidc_configuration () {
+	if (!$jwks) {
+		my $discovery_endpoint = $oidc_options{oidc_discovery_url};
+
+		$log->info('Original OIDC configuration', {discovery_endpoint => $discovery_endpoint})
+			if $log->is_info();
+
+		my $discovery_request = HTTP::Request->new(GET => $discovery_endpoint);
+		my $discovery_response = LWP::UserAgent::Plugin->new->request($discovery_request);
+		unless ($discovery_response->is_success) {
+			$log->info('Unable to load OIDC data from IdP', {response => $discovery_response->content})
+				if $log->is_info();
+			return;
+		}
+
+		$oidc_configuration = decode_json($discovery_response->content);
+		$log->info('got discovery document', {discovery => $oidc_configuration}) if $log->is_info();
+
+		_load_jwks_configuration_to_oidc_options($oidc_configuration->{jwks_uri});
+	}
+
+	return $oidc_configuration;
+}
+
+=head2 _load_jwks_configuration_to_oidc_options( $jwks_uri )
+
+Loads the JWKS from $jwks_uri, and stores it in the $jkw variable.
+
+JWKS aka JSON Web Key Sets are essential to validate access tokens
+https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets
+
+=head3 Arguments
+
+=head4 URI to the JWKS. $jwks_uri
+
+=head3 Return values
+
+None.
+
+=cut
+
+sub _load_jwks_configuration_to_oidc_options ($jwks_uri) {
+	my $jwks_request = HTTP::Request->new(GET => $jwks_uri);
+	my $jwks_response = LWP::UserAgent::Plugin->new->request($jwks_request);
+	unless ($jwks_response->is_success) {
+		$log->info('Unable to load JWKS from IdP', {response => $jwks_response->content}) if $log->is_info();
+		return;
+	}
+
+	$jwks = decode_json($jwks_response->content);
+	$log->info('got JWKS', {jwks => $jwks}) if $log->is_info();
+	return;
+}
+
+=head2 write_auth_deprecated_headers()
+
+Writes the deprecation notice for old authentication sites as HTTP headers.
+
+=head3 Arguments
+
+None.
+
+=head3 Return values
+
+None.
+
+=cut
+
+sub write_auth_deprecated_headers() {
+	if (get_oidc_implementation_level() >= 3) {
+		my $r = Apache2::RequestUtil->request();
+		$r->err_headers_out->set('Deprecation', 'Mon, 01 Apr 2024 00:00:00 GMT');
```

Comment:
TODO: Actual values need to be discussed. Maybe only after the OIDC level is raised to 3

Reply:
I've shifted the dates out a bit for now